pandas/core/internals/array_manager.py

```diff
@@ -352,12 +378,20 @@ def putmask(self, mask, new, align: bool = True):
             align_keys = ["mask"]
             new = extract_array(new, extract_numpy=True)
 
-        return self.apply_with_block(
+        for i in range(len(self.arrays)):
+            if not self._has_no_reference(i):
+                # if being referenced -> perform Copy-on-Write and clear the reference
+                self.arrays[i] = self.arrays[i].copy()
```

Comment:
this is copying more than is strictly necessary (which may be harmless).  to be precise would need to examine the `mask` and exclude columns for which `not mask[i].any()`

Reply:
That's one of the reasons I first wanted to refactor putmask here (https://github.com/pandas-dev/pandas/pull/44396/). But as long as we are using `apply_with_block` here, I am not going to bother looking into such optimization (it would require part of the changes in https://github.com/pandas-dev/pandas/pull/44396 anyway to split up the mask like that)